engines/macventure/sound.cpp

```diff
@@ -0,0 +1,261 @@
+	// TODO: Possible source of bugs, the original just assigns the seek to the scales
+	uint32 scales = stream->pos() + 0xe2;
+	for (uint i = 0; i < repeat; i++) {
+		stream->seek(scales + i * 2, SEEK_SET);
+		uint32 scale = stream->readUint16BE();
+		stream->seek(base + 0xa, SEEK_SET);
+		for (uint j = 0; j < _length; j++) {
+			byte ch = stream->readByte();
+			if (ch & 0x80) {
+				ch -= 0x80;
+				uint32 env= ch * scale;
+				ch = (env >> 8) & 0xff;
+				if (ch & 0x80) ch = 0x7f;
+				ch += 0x80;
+			} else {
+				ch = (ch ^ 0xff) + 1;
+				ch -= 0x80;
+				uint32 env = ch * scale;
+				ch = (env >> 8) & 0xff;
+				if (ch & 0x80) ch = 0x7f;
+				ch += 0x80;
+				ch = (ch ^ 0xff) + 1;
+			}
+			_data.push_back(ch);
+		}
+	}
+}
+
+void SoundAsset::decode18(Common::SeekableReadStream *stream) {
+	warning("Decode sound 0x18 untested");
+	Common::Array<byte> wavtable;
+	stream->seek(0x252, SEEK_SET);
+	for (uint i = 0; i < 16; i++) {
+		wavtable.push_back(stream->readByte());
+	}
+	_length = stream->readUint32BE() * 2;
+	//Unused
+	stream->readUint16BE();
+	_frequency = (stream->readUint32BE() * 22100 / 0x10000) | 0;
+	byte ch = 0;
+	for (uint i = 0; i < _length; i++) {
+		if (i & 1) ch >>= 4;
+		else ch = stream->readByte();
+		_data.push_back(wavtable[ch & 0xf]);
+	}
+}
+
+void SoundAsset::decode1a(Common::SeekableReadStream *stream) {
+	warning("Decode sound 0x1a untested");
+	Common::Array<byte> wavtable;
+	stream->seek(0x220, SEEK_SET);
+	for (uint i = 0; i < 16; i++) {
+		wavtable.push_back(stream->readByte());
+	}
+	_length = stream->readUint32BE();
+	//Unused
+	stream->readUint16BE();
+	_frequency = (stream->readUint32BE() * 22100 / 0x10000) | 0;
+	byte ch = 0;
+	for (uint i = 0; i < _length; i++) {
+		if (i & 1) ch >>= 4;
+		else ch = stream->readByte();
+		_data.push_back(wavtable[ch & 0xf]);
+	}
+}
+
+void SoundAsset::decode44(Common::SeekableReadStream *stream) {
+	stream->seek(0x5e, SEEK_SET);
+	_length = stream->readUint32BE();
+	_frequency = (stream->readUint32BE() * 22100 / 0x10000) | 0;
+	for (uint i = 0; i < _length; i++) {
+		_data.push_back(stream->readByte());
+	}
+}
+
+void SoundAsset::decode78(Common::SeekableReadStream *stream) {
+	Common::Array<byte> wavtable;
+	stream->seek(0xba, SEEK_SET);
+	for (uint i = 0; i < 16; i++) {
+		wavtable.push_back(stream->readByte());
+	}
+	//Unused
+	stream->readUint32BE();
+	_length = stream->readUint32BE();
+	_frequency = (stream->readUint32BE() * 22100 / 0x10000) | 0;
+	byte ch = 0;
+	for (uint i = 0; i < _length; i++) {
+		if (i & 1) ch <<= 4;
+		else ch = stream->readByte();
+		_data.push_back(wavtable[(ch >> 4) & 0xf]);
+	}
+}
+
+void SoundAsset::decode7e(Common::SeekableReadStream *stream) {
+	Common::Array<byte> wavtable;
+	stream->seek(0xc2, SEEK_SET);
+	for (uint i = 0; i < 16; i++) {
+		wavtable.push_back(stream->readByte());
+	}
+	//Unused
+	stream->readUint32BE();
+	_length = stream->readUint32BE();
+	_frequency = (stream->readUint32BE() * 22100 / 0x10000) | 0;
```

Comment:
That's a javascript thing to turn Infinity (caused by divisions by 0) into 0 I think. They'll all need to be changed.